src/gluonts/model/forecast_generator.py

```diff
@@ -102,142 +120,113 @@ def make_distribution_forecast(distr, *args, **kwargs) -> Forecast:
-                )
-            assert i + 1 == len(batch[FieldName.FORECAST_START])
-
-
-class DistributionForecastGenerator(ForecastGenerator):
-    @validated()
-    def __init__(self, distr_output) -> None:
-        self.distr_output = distr_output
-
-    def __call__(
-        self,
-        inference_data_loader: DataLoader,
-        prediction_net,
-        input_names: List[str],
-        output_transform: Optional[OutputTransform],
-        num_samples: Optional[int],
-        **kwargs
-    ) -> Iterator[Forecast]:
-        for batch in inference_data_loader:
-            inputs = [batch[k] for k in input_names]
-            outputs = prediction_net(*inputs)
-
-            if output_transform:
-                log_once(OUTPUT_TRANSFORM_NOT_SUPPORTED_MSG)
-            if num_samples:
-                log_once(NOT_SAMPLE_BASED_MSG)
-
-            distributions = [
-                self.distr_output.distribution(*u)
-                for u in recursively_zip_arrays(outputs)
-            ]
-
-            i = -1
-            for i, distr in enumerate(distributions):
-                yield make_distribution_forecast(
-                    distr,
-                    start_date=batch[FieldName.FORECAST_START][i],
-                    item_id=batch[FieldName.ITEM_ID][i]
-                    if FieldName.ITEM_ID in batch
-                    else None,
-                    info=batch["info"][i] if "info" in batch else None,
-                )
-            assert i + 1 == len(batch[FieldName.FORECAST_START])
+class ForecastBatch:
+    @property
+    def batch_size(self) -> int:
+        raise NotImplementedError
+
+    @property
+    def mean(self) -> np.ndarray:
+        raise NotImplementedError
+
+    def quantile(self, q: Union[float, str]) -> np.ndarray:
+        raise NotImplementedError
+
+
+@dataclass
+class DistributionForecastBatch(ForecastBatch):
+    start_date: list
+    item_id: Optional[list]
+    info: Optional[list]
+    distr_output: Any  # TODO fix
+    distr_args: list
+    distr: Type = field(init=False)
+
+    def __post_init__(self):
+        self.distr = self.distr_output.distribution(*self.distr_args)
+
+    def __iter__(self) -> Iterator[Forecast]:  # TODO fix
+        distributions = [
+            self.distr_output.distribution(*u)
+            for u in recursively_zip_arrays(self.distr_args)
+        ]
+
+        for i, distr in enumerate(distributions):
+            yield make_distribution_forecast(
+                distr,
+                start_date=self.start_date[i],
+                item_id=self.item_id[i] if self.item_id is not None else None,
+                info=self.info[i] if self.info is not None else None,
+            )
+
+    @property
+    def mean(self) -> np.ndarray:
+        return to_numpy(self.distr.mean())
+
+    def quantile(self, q: Union[float, str]) -> np.ndarray:
+        return to_numpy(self.distr.quantile(q))
+
+
+@dataclass
+class SampleForecastBatch(ForecastBatch):
+    start_date: list
```

Comment:
```suggestion
    start: list
```

or maybe `starts`?

And what is the type -- `pd.Period`?